### PR TITLE
Refactor conditional detr 3.x

### PR DESCRIPTION
**Pull request by LYMDLUT**
This PR aims to refactor Conditional DETR in  MMDetection 3.0!

I'm working together with @jshilong, @Li-Qingyun, and @KeiChiTse  to refactor the DETR-like algorithms in MMDetection 3.0, using a more lightweight and easy-to-read code style.

feel free to contact me!

### CODE REVIEW

**Comment by Li-Qingyun** [comment]
Hi, Yiming.
The target branch of your PR may be `refactor-detr`?
You can check and click Edit to change it.

File: mmdet/models/layers/conditional_transformer.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        self.post_norm = build_norm_layer(self.post_norm_cfg,self.embed_dims)[1]
+        #####################################################
+        d_model = 256
+        self.query_scale = MLP(d_model, d_model, d_model, 2)
+        self.ref_point_head = MLP(d_model, d_model, 2, 2)
+        for layer_id in range(self.num_layers - 1):
+            self.layers[layer_id + 1].cross_attn.qpos_proj = None
+        #####################################################
+    def forward(self, query, *args, **kwargs):
+        reference_points_before_sigmoid = self.ref_point_head(kwargs['query_pos'])  # [num_queries, batch_size, 2]
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:62]: I think it is best not to index variables from `kwargs`.

File: mmdet/models/layers/conditional_transformer.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            kwargs['query_sine_embed'] = query_sine_embed
+            kwargs['is_first'] = (layer_id == 0)
+            query = layer(query, *args, **kwargs)
+            if self.return_intermediate:
+                intermediate.append(self.post_norm(query))
+
+        if self.return_intermediate:
+            return [torch.stack(intermediate), reference_points]
+
+        return [query, reference_points]
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:84]: why is here a list?

File: mmdet/models/layers/conditional_transformer.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+
+class ConditionalDetrTransformerDecoderLayer(DetrTransformerDecoderLayer):
+
+    def _init_layers(self):
+        self.self_attn = ConditionalAttention(**self.self_attn_cfg)
+        self.cross_attn =ConditionalAttention(**self.cross_attn_cfg)
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:91]: space

File: mmdet/models/layers/conditional_transformer.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        assert key.size(0) == value.size(0), \
+            f'{"key, value must have the same sequence length"}'
+        assert query.size(1) == key.size(1) == value.size(1), \
+            f'{"batch size must be equal for query, key, value"}'
+        assert query.size(2) == key.size(2), \
+            f'{"q_dims, k_dims must be equal"}'
+        assert value.size(2) == self.embed_dims, \
+            f'{"v_dims must be equal to embed_dims"}'
+
+        tgt_len, bsz, hidden_dims = query.size()
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:185]: remember to modify all about batch_size   bs/bsz --> batch_size

File: mmdet/models/layers/conditional_transformer.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    def forward_attn(self, query, key, value, attn_mask, key_padding_mask,
+                     need_weights):
+        assert key.size(0) == value.size(0), \
+            f'{"key, value must have the same sequence length"}'
+        assert query.size(1) == key.size(1) == value.size(1), \
+            f'{"batch size must be equal for query, key, value"}'
+        assert query.size(2) == key.size(2), \
+            f'{"q_dims, k_dims must be equal"}'
+        assert value.size(2) == self.embed_dims, \
+            f'{"v_dims must be equal to embed_dims"}'
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:183]: It looks vaguely familiar.
but why the prompt information in the assert message is deleted?
```Python
assert query.size(2) == key.size(2), \
    f'The q_dim {query.size(2)} must be equal to the k_dim' \
    f' {key.size(2)}'
assert value.size(2) == self.out_dims, \
    f'The v_dim {value.size(2)} must be equal to the out_dims ' \
    f'{self.out_dims}'
assert key.size(0) == value.size(0), \
    f'The k_len {key.size(0)} must be equal to the v_len ' \
    f'{value.size(0)}'
assert query.size(1) == key.size(1) == value.size(1), \
    f'The batch size (2nd) of q, k, v must be equal, ' \
    f'but got {query.size(1)}(q), {key.size(1)}(k), ' \
    f'{value.size(1)}(v)'
```

File: mmdet/models/layers/conditional_transformer.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+        if key_padding_mask is not None:
+            attn_output_weights = attn_output_weights.view(
+                bsz, self.num_heads, tgt_len, src_len)
+            attn_output_weights = attn_output_weights.masked_fill(
+                key_padding_mask.unsqueeze(1).unsqueeze(2),
+                float('-inf'),
+            )
+            attn_output_weights = attn_output_weights.view(
+                bsz * self.num_heads, tgt_len, src_len)
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:265]: We should discuss whether it's necessary to keep so much complex logic about masks in ConditionalAttention.

**Comment by LYMDLUT** [comment]
> Hi, Yiming. The target branch of your PR may be `refactor-detr`? You can check and click Edit to change it.

Thank you，I have changed it

File: mmdet/models/layers/conditional_transformer.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+        attn_output_weights = F.softmax(attn_output_weights, dim=-1)
```
Comment (Li-Qingyun) [inline mmdet/models/layers/conditional_transformer.py:267]: We should discuss whether to leave an interface here to support dab in ConditionalAttention.

**Review by Li-Qingyun — COMMENTED** [review]
Great job !

File: mmdet/models/dense_heads/conditional_detr_head.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import torch.nn as nn
+from torch import Tensor
+
+from mmdet.registry import MODELS
+from .detr_head import DETRHead
+from mmdet.models.layers.transformer import inverse_sigmoid
+
+def bias_init_with_prob(prior_prob: float) -> float:
+    """initialize conv/fc bias value according to a given probability value."""
+    import numpy as np
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/models/dense_heads/conditional_detr_head.py:14]: Please use bias_init_with_prob in the mmcv, you can refer to other dense_head

File: mmdet/models/dense_heads/conditional_detr_head.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    Args:
+        TODO
+    """
+
+    def init_weights(self):
+        """Initialize weights of the transformer head."""
+        # The initialization for transformer is important
+        if self.loss_cls.use_sigmoid:
+            bias_init = bias_init_with_prob(0.01)
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/models/dense_heads/conditional_detr_head.py:33]: There is same function in mmcv

File: mmdet/models/dense_heads/conditional_detr_head.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+            - all_cls_scores (Tensor): Outputs from the classification head, \
+            shape [nb_dec, bs, num_query, cls_out_channels]. Note \
+            cls_out_channels should includes background.
+            - all_bbox_preds (Tensor): Sigmoid outputs from the regression \
+            head with normalized coordinate format (cx, cy, w, h). \
+            Shape [nb_dec, bs, num_query, 4].
+        """
+        out_dec, reference = outs_dec[0], outs_dec[1]
+        ##################
```
Comment (jshilong) [inline mmdet/models/dense_heads/conditional_detr_head.py:52]: remove

File: mmdet/models/dense_heads/conditional_detr_head.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        reference_before_sigmoid = inverse_sigmoid(reference)
+        outputs_coords = []
+        for lvl in range(out_dec.shape[0]):
+            tmp = self.fc_reg(self.activate(self.reg_ffn(out_dec[lvl])))
+            tmp[..., :2] += reference_before_sigmoid
+            outputs_coord = tmp.sigmoid()
+            outputs_coords.append(outputs_coord)
+        outputs_coord = torch.stack(outputs_coords)
+        ########################
```
Comment (jshilong) [inline mmdet/models/dense_heads/conditional_detr_head.py:61]: remove

File: mmdet/models/layers/conditional_transformer.py (hunk 7 above)
Comment (LYMDLUT) [reply mmdet/models/layers/conditional_transformer.py:267]: I think we can move it to layers/transfomer.py, and rewrite F.softmax function

File: mmdet/models/dense_heads/conditional_detr_head.py (hunk 11 above)
Comment (LYMDLUT) [reply mmdet/models/dense_heads/conditional_detr_head.py:61]: ok

File: configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco_class80_focal.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                                    (576, 1333), (608, 1333), (640, 1333),
+                                    (672, 1333), (704, 1333), (736, 1333),
+                                    (768, 1333), (800, 1333)],
+                            keep_ratio=True)
+                    ]]),
+    dict(type='PackDetInputs')
+]
+
+# if categories is 80, set class80=True; if categories is 91, set calss80=80
+class80 = True
```
[the remark this answers is not in the document]
Comment (LYMDLUT) [reply configs/conditional_detr/conditional_detr_r50_8xb2-50e_coco_class80_focal.py:117]: ok,I will immidiately delete the support for 91 classes as the fomal code

File: mmdet/models/dense_heads/deformable_detr_head.py [hunk 13]
```diff
@@ -96,63 +89,12 @@ def init_weights(self) -> None:
             for m in self.reg_branches:
                 nn.init.constant_(m[-1].bias.data[2:], 0.0)
 
-    def forward(self, x: Tuple[Tensor],
-                batch_img_metas: List[dict]) -> Tuple[Tensor, ...]:
+    def forward(self, hs: Tensor, init_reference,
```
Comment (jshilong) [inline mmdet/models/dense_heads/deformable_detr_head.py:92]: @Li-Qingyun we should rename this `hs` to a more reasonable name

Comment (Li-Qingyun) [reply mmdet/models/dense_heads/deformable_detr_head.py:92]: hidden_states?

File: mmdet/models/detectors/conditional_detr.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+class ConditionalDETR(DETR):
+    r"""Implementation of `DETR: End-to-End Object Detection with
+    Transformers <https://arxiv.org/pdf/2005.12872>`_"""
+
+    def _init_layers(self) -> None:
+        self.positional_encoding = SinePositionalEncoding(
+            **self.positional_encoding_cfg)
+        self.encoder = DetrTransformerEncoder(**self.encoder_cfg)
+        # NOTE only the line below has beeen changed
+        self.decoder = ConditionalDetrTransformerDecoder(**self.decoder_cfg)
```
[the remark this answers is not in the document]
Comment (LYMDLUT) [reply mmdet/models/detectors/conditional_detr.py:21]: Good idea! Most of our work just change the name of Encoder and Decoder,and the other modules are same

File: mmdet/models/layers/conditional_transformer.py (hunk 1 above)
Comment (LYMDLUT) [reply mmdet/models/layers/conditional_transformer.py:61]: I have fix it

File: mmdet/models/layers/conditional_transformer.py (hunk 2 above)
Comment (LYMDLUT) [reply mmdet/models/layers/conditional_transformer.py:83]: Beacause I do not want to reload the decoder function' forward_transfomer 'in detr.py.and I also have changed    out_dec=out_dec.transpose(1,2) to out_dec[0] = out_dec[0].transpose(1, 2) to supoort this

Comment (LYMDLUT) [reply mmdet/models/layers/conditional_transformer.py:83]: https://github.com/Li-Qingyun/mmdetection/blob/eb3a9d17e72aad1814719c420e13858fd76649a9/mmdet/models/detectors/detr.py#L110

Comment (Li-Qingyun) [reply mmdet/models/layers/conditional_transformer.py:82]: I have not thought the modification in 3581a00fab18db7f4e09c01d11bcf559321f5c93 is suitable.
We can have a discussion at our next meeting.

**Comment by jshilong** [comment]
Please merge the latest refactor_detr

**Comment by ZwwWayne** [comment]
Closed as it has been solved in #9405
